app/javascript/crayons/MultiSelectAutocomplete/MultiSelectAutocomplete.jsx

```diff
@@ -0,0 +1,388 @@
+  };
+
+  const handleKeyDown = (e) => {
+    const { selectionStart, value: currentValue } = inputRef.current;
+
+    switch (e.key) {
+      case KEYS.DOWN:
+        e.preventDefault();
+
+        if (
+          activeDescendentIndex !== null &&
+          activeDescendentIndex < suggestions.length - 1
+        ) {
+          dispatch({
+            type: 'setActiveDescendentIndex',
+            payload: activeDescendentIndex + 1,
+          });
+        } else {
+          dispatch({ type: 'setActiveDescendentIndex', payload: 0 });
+        }
+        break;
+      case KEYS.UP:
+        e.preventDefault();
+
+        dispatch({
+          type: 'setActiveDescendentIndex',
+          payload:
+            activeDescendentIndex >= 1
+              ? activeDescendentIndex - 1
+              : suggestions.length - 1,
+        });
+
+        break;
+      case KEYS.ENTER:
+        e.preventDefault();
+        if (activeDescendentIndex !== null) {
+          selectItem({ selectedItem: suggestions[activeDescendentIndex] });
+        }
+        break;
+      case KEYS.ESCAPE:
+        e.preventDefault();
+        // Clear the input and suggestions
+        clearInput();
+        break;
+      case KEYS.COMMA:
+      case KEYS.SPACE:
+        e.preventDefault();
+        // Accept whatever is in the input before the comma or space.
+        // If any text remains after the comma or space, the edit will continue separately
+        if (currentValue !== '') {
+          selectItem({
+            selectedItem: currentValue.slice(0, selectionStart),
+            nextInputValue: currentValue.slice(selectionStart),
+          });
+        }
+        break;
+      case KEYS.DELETE:
+        if (currentValue === '') {
+          e.preventDefault();
+          editPreviousSelectionIfExists();
+        }
+        break;
+      default:
+        if (!ALLOWED_CHARS_REGEX.test(e.key)) {
+          e.preventDefault();
+        }
+    }
+  };
+
+  // If there is a previous selection, then pop it into edit mode
+  const editPreviousSelectionIfExists = () => {
+    if (selectedItems.length > 0 && inputPosition !== 0) {
+      const nextEditIndex =
+        inputPosition !== null ? inputPosition - 1 : selectedItems.length - 1;
+
+      const item = selectedItems[nextEditIndex];
+      deselectItem(item);
+      enterEditState(item, nextEditIndex);
+    }
+  };
+
+  const selectItem = ({
+    selectedItem,
+    nextInputValue = '',
+    focusInput = true,
+  }) => {
+    // If a user has manually typed an item already selected, reset
+    if (selectedItems.includes(selectedItem)) {
+      clearInput();
+      return;
+    }
+
+    // If an item was edited, we want to keep it in the same position in the list
+    const insertIndex =
+      inputPosition !== null ? inputPosition : selectedItems.length;
+    const newSelections = [
+      ...selectedItems.slice(0, insertIndex),
+      selectedItem,
+      ...selectedItems.slice(insertIndex),
+    ];
```

Comment:
Refactor (blocking): [Array.prototype.slice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice) returns a shallow copy. Spreading the array before the splice is redundant.

```suggestion
    const newSelections = [
      selectedItems.slice(0, insertIndex),
      selectedItem,
      selectedItems.slice(insertIndex),
    ];
```

Reply:
Hmm, I'm spreading the _result_ of the `slice()` - if I don't spread here I end up with `[[first sliced group], selectedItem, [second sliced group]]`